precompiles/dispatch-lockdrop/src/lib.rs

```diff
@@ -0,0 +1,159 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Astar is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Astar. If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate alloc;
+
+use alloc::format;
+use core::marker::PhantomData;
+use fp_evm::{ExitError, PrecompileFailure, PrecompileHandle};
+use frame_support::pallet_prelude::IsType;
+use frame_support::{codec::DecodeLimit as _, dispatch::Pays, traits::Get};
+use frame_support::{
+    dispatch::{Dispatchable, GetDispatchInfo, PostDispatchInfo},
+    traits::ConstU32,
+};
+use frame_system::Config;
+use pallet_evm::GasWeightMapping;
+use pallet_evm_precompile_dispatch::DispatchValidateT;
+use precompile_utils::prelude::{revert, BoundedBytes, UnboundedBytes};
+use precompile_utils::EvmResult;
+use sp_core::{crypto::AccountId32, H160, H256};
+use sp_io::hashing::keccak_256;
+use sp_std::vec::Vec;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub const LOG_TARGET: &str = "precompile::dispatch-lockdrop";
+
+// ECDSA PublicKey
+type ECDSAPublic = ConstU32<64>;
+
+// `DecodeLimit` specifies the max depth a call can use when decoding, as unbounded depth
+// can be used to overflow the stack.
+// Default value is 8, which is the same as in XCM call decoding.
+pub struct DispatchLockdrop<Runtime, DispatchValidator, DecodeLimit = ConstU32<8>>(
+    PhantomData<(Runtime, DispatchValidator, DecodeLimit)>,
+);
+
+#[precompile_utils::precompile]
+impl<Runtime, DispatchValidator, DecodeLimit>
+    DispatchLockdrop<Runtime, DispatchValidator, DecodeLimit>
+where
+    Runtime: pallet_evm::Config,
+    <Runtime::RuntimeCall as Dispatchable>::RuntimeOrigin: From<Option<Runtime::AccountId>>,
+    Runtime::RuntimeCall: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+    <Runtime as Config>::AccountId: IsType<AccountId32>,
+    <Runtime as Config>::AccountId: From<[u8; 32]>,
+    DispatchValidator:
+        DispatchValidateT<<Runtime as Config>::AccountId, <Runtime as Config>::RuntimeCall>,
+    DecodeLimit: Get<u32>,
+{
+    #[precompile::public("dispatch_lockdrop_call(bytes,bytes)")]
+    fn dispatch_lockdrop_call(
+        handle: &mut impl PrecompileHandle,
+        call: UnboundedBytes,
+        pubkey: BoundedBytes<ECDSAPublic>,
+    ) -> EvmResult<bool> {
+        log::trace!(
+            target: LOG_TARGET,
+            "raw arguments: call: {:?}, pubkey: {:?}",
+            call,
+            pubkey
+        );
+
+        let caller: H160 = handle.context().caller.into();
+        let input: Vec<u8> = call.into();
+
+        // 1. Decode the call
```

Comment:
I'm not sure how expensive all these operations are (ignoring the call dispatch), but I'd suggest to add some static cost at the beginning, that's always charged. Sort of a deterrent to prevent spam attack and continued use of this precompile.

Correct me if I'm wrong, but we just want users to get their funds out and continue using "normal" ways of interacting with the chain.

Reply:
Ok, I think I understand where we're misunderstanding each other.

I see two problems:
1. Double charging
2. Incorrect charging

### Double charging
The first one, double charging, happens because as you outlined your steps:
2. Charge external cost
3.1. Charge actual cost
3.2. Refund the initial charged weight

After the 2nd step, you have double charged.
It's true that you refund it later, but at that point the total charged amount might have exceeded what would otherwise be good enough good limit if it weren't for _temporary_ double charging.

### Incorrect charging

Later on, when you call:
```Rust
let actual_weight = post_info.actual_weight.unwrap_or(info.weight);
let cost = Runtime::GasWeightMapping::weight_to_gas(actual_weight);
handle.record_cost(cost)?;
```

you will actually ignore the PoV part of the weight.
```Rust
fn weight_to_gas(weight: Weight) -> u64 {
    weight.div(T::WeightPerGas::get().ref_time()).ref_time()
}
```

This is the approach used by `try_dispatch`:
```Rust
    pub fn try_dispatch<Call>(
        handle: &mut impl PrecompileHandle,
        origin: <Runtime::RuntimeCall as Dispatchable>::RuntimeOrigin,
        call: Call,
    ) -> Result<PostDispatchInfo, TryDispatchError>
    where
        Runtime::RuntimeCall: From<Call>,
    {
        let call = Runtime::RuntimeCall::from(call);
        let dispatch_info = call.get_dispatch_info();

        Self::record_weight_v2_cost(handle, dispatch_info.weight).map_err(TryDispatchError::Evm)?;

        // Dispatch call.
        // It may be possible to not record gas cost if the call returns Pays::No.
        // However while Substrate handle checking weight while not making the sender pay for it,
        // the EVM doesn't. It seems this safer to always record the costs to avoid unmetered
        // computations.
        let post_dispatch_info = using_precompile_handle(handle, || call.dispatch(origin))
            .map_err(|e| TryDispatchError::Substrate(e.error))?;

        Self::refund_weight_v2_cost(
            handle,
            dispatch_info.weight,
            post_dispatch_info.actual_weight,
        )
        .map_err(TryDispatchError::Evm)?;

        Ok(post_dispatch_info)
    }
}
```

### Suggestion

1. At the very beginning of the call you charge the hardcoded/fixed value to cover for all the decoding, hashing, etc. This is purely computation (ref_time, gas), no PoV is involved.
2. When you get to the part when it's time to dispatch the call, use the approach from `try_dispatch`, which records both weight components.

Reply:
1. **Double Charging:** I don't see any purpose for that `record_cost` call, it can be removed completely as it doesn't really make sense to double charge ref_time.
2. **Incorrect charging**: We should probably charge a hefty fee from users using this precompile, as Dino said _we just want users to get their funds out and continue using "normal" ways of interacting with the chain_,. Charging a large fee will surely discourage people using it for other purposes.